domain/resource/state/resource.go

```diff
@@ -407,173 +419,568 @@ WHERE uuid = $resourceIdentity.uuid`,
-WHERE unit_resource.resource_uuid = $unitResource.resource_uuid 
-AND unit_resource.unit_uuid = $unitResource.unit_uuid`
-	checkUnitResourceStmt, err := st.Prepare(checkUnitResourceQuery, unitResourceInput)
+
+	kind, err := charmresource.ParseType(resKind.Name)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Errorf("parsing resource kind: %w", err)
 	}
+	return kind, err
+}
 
-	// Prepare statement to check that UnitUUID is valid UUID.
-	unitUUID := unitNameAndUUID{UnitUUID: config.UnitUUID}
-	checkValidUnitQuery := `
-SELECT &unitNameAndUUID.uuid 
-FROM unit 
-WHERE uuid = $unitNameAndUUID.uuid`
-	checkValidUnitStmt, err := st.Prepare(checkValidUnitQuery, unitUUID)
+// insertFileResource checks that the storage ID corresponds to stored object
+// store metadata and then records that the resource is stored at the provided
+// storage ID.
+func (st *State) insertFileResource(
+	ctx context.Context,
+	tx *sqlair.TX,
+	resourceUUID coreresource.UUID,
+	storageID coreresourcestore.ID,
+) error {
+	// Get the object store UUID of the stored resource blob.
+	if !storageID.IsObjectStoreUUID() {
+		return errors.Errorf("cannot insert file resource that is not stored in the object store")
+	}
+	uuid, err := storageID.ObjectStoreUUID()
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return errors.Capture(err)
 	}
 
-	// Prepare statement to check that resourceID is valid UUID.
-	resourceUUID := resourceIdentity{UUID: config.ResourceUUID.String()}
-	checkValidResourceQuery := `
-SELECT &resourceIdentity.uuid
-FROM resource
-WHERE uuid = $resourceIdentity.uuid`
-	checkValidResourceStmt, err := st.Prepare(checkValidResourceQuery, resourceUUID)
+	// Check the resource blob is stored in the object store.
+	storedResource := storedFileResource{
+		ResourceUUID:    resourceUUID.String(),
+		ObjectStoreUUID: uuid.String(),
+	}
+	checkObjectStoreMetadataStmt, err := st.Prepare(`
+SELECT uuid AS &storedFileResource.store_uuid
+FROM object_store_metadata
+WHERE uuid = $storedFileResource.store_uuid
+`, storedResource)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return errors.Capture(err)
 	}
 
-	// Prepare statement to verify if the application resource is already
-	// retrieved.
+	err = tx.Query(ctx, checkObjectStoreMetadataStmt, storedResource).Get(&storedResource)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf("checking object store for resource %s: %w", resourceUUID, resourceerrors.StoredResourceNotFound)
+	} else if err != nil {
+		return errors.Errorf("checking object store for resource %s: %w", resourceUUID, err)
+	}
+
+	// Check if the resource has already been stored.
+	checkResourceFileStoreStmt, err := st.Prepare(`
+SELECT &storedFileResource.*
+FROM   resource_file_store
+WHERE  resource_uuid = $storedFileResource.resource_uuid
+`, storedResource)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
```

Comment:
I think it would be possible, but I would rather keep them separate. They're checking for two different things here and have different errors for each if the check fails. In the first query we expect to find the thing, in the second we don't. I might be possible to make them a single query but code a lot more confusing as it would probably involve a left join and check for the joined value being empty rather than relying on `sqlair.ErrNoRows` as we do now.